core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java

```diff
@@ -132,29 +132,32 @@ final class RampingUpEndpointWeightSelector extends AbstractEndpointSelector {
 
         @Nullable
         private Set<EndpointAndStep> unhandledNewEndpoints;
+        private final AtomicBoolean initialized = new AtomicBoolean();
 
         RampingUpEndpointWeightSelector(EndpointGroup endpointGroup, EventExecutor executor) {
             super(endpointGroup);
             this.executor = executor;
 
-            final AtomicBoolean initialized = new AtomicBoolean();
-            endpointGroup.addListener(newEndpoints -> {
-                if (initialized.compareAndSet(false, true)) {
-                    final List<Endpoint> dedupEndpoints =
-                            new ArrayList<>(deduplicateEndpoints(newEndpoints).values());
-                    endpointSelector = new WeightedRandomDistributionEndpointSelector(dedupEndpoints);
-                    endpointsFinishedRampingUp.addAll(dedupEndpoints);
-                } else {
-                    // Use the executor so the order of endpoints change is guaranteed.
-                    executor.execute(() -> updateEndpoints(newEndpoints));
-                }
-            }, true);
-
+            endpointGroup.whenReady().thenAccept(unused -> maybeUpdateEndpoints(endpointGroup.endpoints()));
             if (endpointGroup instanceof ListenableAsyncCloseable) {
                 ((ListenableAsyncCloseable) endpointGroup).whenClosed().thenRunAsync(this::close, executor);
             }
         }
 
+        private void maybeUpdateEndpoints(List<Endpoint> endpoints) {
+            if (initialized.compareAndSet(false, true)) {
+                final List<Endpoint> dedupEndpoints =
+                        new ArrayList<>(deduplicateEndpoints(endpoints).values());
+                endpointSelector = new WeightedRandomDistributionEndpointSelector(dedupEndpoints);
+                endpointsFinishedRampingUp.addAll(dedupEndpoints);
+
+                group().addListener(this::maybeUpdateEndpoints);
```

Comment:
Can't this be
```suggestion
                group().addListener(this::updateEndpoints);
```
because `initialized` never changes back to `false`?

Reply:
Oops... that's my fault.